Ship parameter rework.

CardamomCake opened the pull request:
This PR definitely deserves explanation, and I'l try to do my best to keep it concise. Some of my solutions might have had a more straightforward solution, but there ware a lot of limitations and edge cases of dotnet, and the json parser that finding a way to work around all of those at all was difficult. 

Firstly, I removed the object containing the default parameters, and replaced it with a function. This I did because I ran into a problem that when I loaded a level/game where I changed the parameters during the loading process, the defaults would end up changed too. Despite them being "read only", they would somehow end up overwritten, breaking the check on if the race is valid. The solution for this was to make it a function instead, since this is immune from getting overwritten. I also changed every instance of the object getting called so the game is not broken. 

I gave everything a default value so that the json parser could understand the difference between a parameter that is missing, and a parameter that is set to zero. This is also why all the default values are constants now, otherwise the compiler would not allow this. 

I have added a list of the string names of all the ship parameters, this serves two functions. One being a list of names which can be searched up in the JSON structure, and that can be iterated against. Also it functions as the order of operations for the hash string generator, because JSON order can't be trusted (and you can't apparently iterate against a list of constants), but that is not included in this PR.

jukibom commented:
okay so a couple of things here...

> I ran into a problem that when I loaded a level/game where I changed the parameters during the loading process, the defaults would end up changed too.

That's my bad, and probably shows my inexperience when initially creating this! It's a classic pass-by-reference problem where the defaults object is marked `static readonly` but the contents are not and the assignment is leading to _referencing_ the defaults rather than copying them. Most likely swapping the get to a function and returning a new object would solve this problem (the allocation really isn't a big deal, it can be cached where it's needed) 

```
        public static ShipParameters Defaults() {
            return new() {
                mass = 1100f,
                // etc
``` 

I see the problem this is trying to solve though which is 
* getting the defaults externally, safely
* defining the defaults in one place only
* providing defaults to the JSON parser

and I'm not certain I have a better suggestion on that soo...

> I have added a list of the string names of all the ship parameters, this serves two functions. One being a list of names which can be searched up in the JSON structure, and that can be iterated against. 

This I don't like as it defines the properties multiple times, as strings. We could just use something like

```
foreach (var fieldInfo in typeof(ShipParameters).GetFields(BindingFlags.Public)) {
    var parameter = fieldInfo.Name;
    if (!parameters.ContainsKey(parameter))
        parameters.Add(parameter, defaults[parameter]);
}
```

> Also it functions as the order of operations for the hash string generator, because JSON order can't be trusted (and you can't apparently iterate against a list of constants), but that is not included in this PR.

Worth asking what the hash string generator is actually used for in this case. I'm pretty sure it's only used to generate a filename, not verify the file. In any case, you can enforce the json ordering with `[JsonProperty(Order = 1)]`.

CardamomCake commented:
You were right that the name list is actually not necessary, I have added order properties for each variable, and removed the string list now. 

I understand that you like to keep the default values defined in the Defaults() function, however this causes problems.

1. Newtonsoft requires either variables that can be set to null, or variables that have a default value to distinguish between missing and zero indexes. 
2. setting the flight parameters to be allowed null, causes problems in the flight engine. So I don't think this is a good option
3. Default values for the parameters require the parameters to be known at compile time and have to be constants. Defining them in the Defaults() function will be redundant as they must also be defined as constants.

If you know a better way, I would love to hear it.

jukibom commented:
Apologies on the delay on this one! I wanted to check it out myself and poke around a bit. I noticed there's no real reason to do the iteration and assignment in the `FromJsonString` with the DefaultValue attributes on the base values (it works functionally the same; missing var => use the default, valid but zero var => use zero) so unless there was anything I'm missing there I'm going to merge this and make that change on mainline.
